QA Runs Initial Backend Implementation

ikreymer opened the pull request:
Supports running QA Runs via the QA API:

<img width="1408" alt="Screenshot 2024-03-10 at 6 20 36 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/1015759/4785568a-d52c-4110-8602-8b506612e38f">


Builds on top of the `issue-1498-crawl-qa-backend-support` branch, fixes #1498

Also requires the latest Browsertrix Crawler from webrecorder/browsertrix-crawler#469 branch.

Notable changes:

- Various crawl db operations can be performed on either the crawl or `qa.` object, and core crawl fields have been moved to CoreCrawlable.

- While running,`QARun` data stored in a single `qa` object, while finished qa runs are added to `qaFinished` dictionary on the Crawl. The QA list API returns data from the finished list, sorted by most recent first.

- Includes additional type fixes / type safety, especially around BaseCrawl / Crawl / UploadedCrawl functionality, also creating specific get_upload(), get_basecrawl(), get_crawl() getters for internal use and get_crawl_out() for API

Still need:
- [x] Tests!
- [x] Determining how active QA data is returned (separate API or part of finished list, or just part of crawl).
- [x] Filter page data by a single QA run

tw4l committed:
Start adding backend support to start and stop qa crawl jobs

tw4l committed:
Fixups

tw4l committed:
WIP: Add replay.json endpoint for QA crawls

tw4l committed:
Add API endpoint to return QACrawls associated with a crawl

tw4l committed:
Flesh out method docstring

tw4l committed:
Fixups

tw4l committed:
Minor touchups

tw4l committed:
Use qa-crawlid-ts naming scheme for QA crawls

ikreymer committed:
Merge branch 'main' into issue-1498-crawl-qa-backend-support

ikreymer committed:
qa run work, unifying interface

ikreymer committed:
more refactoring of apis for qa

ikreymer committed:
update operator + crawl_ops calls to pass crawl_id + source_crawl_id
and treat as 'crawl_or_qa_run_id' and 'qa_source_crawl_id' in crawl ops

ikreymer committed:
store qa compare data in unified PageQACompare from page_dict

ikreymer committed:
fixes

ikreymer committed:
refactor basecrawls / crawls / uploads getters:
- rename original get_crawl -> get_crawl_out as its used for api response
- basecrawls has get_base_crawl and get_crawl_raw and get_crawl_out
- crawls has get_crawl
- uploads has get_uplads

ikreymer committed:
various fixes, add separate configmap for qa
qa runs work!

ikreymer committed:
simplify qa model:
- active qa data added to 'qa'
- finished qa later added to 'qa_finished' when crawl finished
- active qa cleared on operator finalize
- default text extract to 'to-warc' for qa support

ikreymer committed:
rename qa_finished -> qaFinished
fix qa replay
additional type fixes, look up Org if missing, needed for resolving presigned URLs

ikreymer committed:
various type fixes to fix tests:
- ensure .from_dict() has proper generic type
- use type Crawl objects for better type safety
- add tags to BaseCrawl
- fix delete_crawls() typing

test: update for page model change

ikreymer committed:
move finished qa run to fininshed in finalizer, perform clear qa + add to finished atomatically
set ttl for qa crawljobs to 0 to finalize more quickly
add shared 'qaCrawlExecSeconds' to crawl to track shared qa exec time usage

ikreymer committed:
page ops: create index on 'crawl_id'

ikreymer committed:
Merge branch 'main' into qa-run

ikreymer committed:
lint fix

tw4l commented on backend/btrixcloud/basecrawls.py:
```diff
@@ -451,12 +433,17 @@ async def _resolve_signed_urls(
+
+                prefix = "files"
+                if qa_run_id:
+                    prefix = f"qa.{qa_run_id}.{prefix}"
```
```suggestion
                    prefix = f"qaFinished.{qa_run_id}.{prefix}"
```

tw4l commented on backend/btrixcloud/crawls.py:
```diff
@@ -735,15 +993,72 @@ async def get_crawl_admin(crawl_id, user: User = Depends(user_dep)):
+        crawl_id: str, qa_run_id: str, org: Organization = Depends(org_crawl_dep)
+    ):
+        # pylint: disable=unused-argument
+        return await ops.delete_crawl_qa_run(crawl_id, qa_run_id)
```
Our other delete endpoints tend to be POSTs who take a list of ids to be able to delete >= 1 resource at a time. Do we want to do that here as well?

ikreymer replied:
Hm, i guess there's actually a mix of POST and DELETE (which should probably standardize), but  making it take a list would be more flexible, so maybe should support that to match the delete crawls API.

ikreymer replied:
Added in: https://github.com/webrecorder/browsertrix-cloud/pull/1586/commits/310780a8a87c00626a22138af3fd4231415c7ef7

tw4l replied:
Yeah, `DELETE` would be more semantically meaningful, but there's inconsistent support for DELETE requests supporting payloads so last round of standardization we went with POST for any endpoints where we might want to delete >1 at a time. You're right, might wanna come back and better standardize this across the board.

tw4l commented on backend/btrixcloud/db.py:
```diff
@@ -210,10 +221,10 @@ def id_str(self):
 
     @classmethod
-    def from_dict(cls, data):
+    def from_dict(cls: Type[T], data: dict) -> T:
```
Nice typing work!

ikreymer replied:
Yeah, this helped find some issues!

tw4l commented on backend/btrixcloud/main.py:
```diff
@@ -126,10 +126,10 @@ def main():
     base_crawl_init = (
         app,
         current_active_user,
+        # to basecrawls
```
What is this comment indicating? Makes sense to move crawl_manager to crawls though.

ikreymer replied:
Oh, that was to indicate that below this, the params are passed as `*args` to BaseCrawls constructor, while the two above are handled directly in init_basecrawls(), init_uploads() and init_crawls(), eg. they have signature:
```
def init_base_crawls_api(app, user_dep, *args):
init_crawls_api(crawl_manager: CrawlManager, app, user_dep, *args):
def init_uploads_api(app, user_dep, *args):
```

and are then called with:
```
init_base_crawls(*base_crawl_init)
init_crawls(crawl_manager, *base_crawl_init)
init_uploads(*base_crawl_init)
```

tw4l commented on backend/btrixcloud/operator/models.py:
```diff
@@ -60,6 +60,17 @@ class CrawlSpec(BaseModel):
     max_crawl_size: int = 0
+    qa_source_crawl_id: str = ""
+
+    @property
+    def db_crawl_id(self) -> str:
+        """return actual crawl_id for db, if qa run"""
+        return self.qa_source_crawl_id or self.id
+
+    @property
+    def is_qa(self) -> bool:
+        """return true if qa run"""
+        return bool(self.qa_source_crawl_id)
```
Nice

tw4l reviewed (commented):
Some initial comments.

Initial testing shows that some pages aren't getting updated, possibly if they redirect to another page.

tw4l commented on backend/btrixcloud/operator/crawls.py:
```diff
@@ -912,7 +951,7 @@ async def increment_pod_exec_time(
 
         if exec_time:
-            await self.org_ops.inc_org_time_stats(oid, exec_time, True)
+            await self.org_ops.inc_org_time_stats(crawl.oid, exec_time, True)
```
Do we want to track monthly QA time at the org level as well?

ikreymer replied:
Currently it should be tracked together with crawl time, but wonder if it should also be separate?
Hm, I guess could add that.

ikreymer replied:
Implemented via: https://github.com/webrecorder/browsertrix-cloud/pull/1586/commits/d61b8d91860eb85e6de57298e305dd4b781157b2
storing in `qausage` and `qaexecTime` (also tracking QA-only usage)

tw4l replied:
Bit of a nitpick, but could we make these `qaUsage` and `qaCrawlExecSeconds` to be more consistent/readable?

ikreymer replied:
Done!

ikreymer committed:
additional pages APIs:
- /qa/<qa_run_id>/pages
- /activeQA

- split Page model into core Page, Page with all qa, and Page with single QA data
- support filtering with qa_run_id, lte / gte, qa_range_field query args

- fix typo in qa stats, use float instead of int!

ikreymer committed:
update test, remove 'qa' from default /pages response

tw4l committed:
Adjust priorities without changing bg-jobs

tw4l committed:
Change qa crawl priority back

tw4l committed:
Allow None for stats for reverse compatibility

tw4l committed:
Add conditional for if base crawl has stats

ikreymer committed:
bg job priority: rename to 'bg-job' (to avoid conflict) and change to -1000

ikreymer committed:
lint: fix conditional for stats being optional

tw4l committed:
Remove duplicate notes field in sort_fields

tw4l committed:
Print log pages data in operator

ikreymer committed:
additional fixes:
- fix qa run stats update while running
- fix qa run pages, add filterBy, gte, gt, lte and lt query args
- include active QA in QA list, update model

ikreymer committed:
page model rename: rename load_state -> loadState, timestamp -> ts to match
pages.jsonl data model

ikreymer committed:
tests: fix page test

ikreymer committed:
initial QA tests!
use 1.1.0-beta.0 crawler for QA support

ikreymer commented:
Per discussion, additional APIs have been added / modified:

.../activeQA - returns only the active QA
.../qa - list includes activeQA data first.
.../qa/{qa_run_id}/pages - filters pages by `qa_run_id`, also supports `filterBy=[screenshotMatch | textMatch]` query arg, and `gt=`, `gte=`, `lt=`, `lte=` query args for filtering pages by specific ranges.

ikreymer committed:
qa: fix qa deletion, add test for deleting qa run, ensuring removed from both crawl and pages

ikreymer committed:
Update backend/btrixcloud/basecrawls.py

Co-authored-by: Tessa Walsh <tessa@bitarchivist.net>

ikreymer committed:
also track qa-only exec time and usage, add qausage and qaexectime to org model

ikreymer committed:
change delete endpoint to use POST, accept a list of qa_run_ids

ikreymer committed:
bump kubernetes-asyncio to 29.0.0

ikreymer committed:
fix qa org keys to be properly cased

ikreymer committed:
Merge branch 'main' into qa-run

ikreymer committed:
additional type fixes

ikreymer committed:
fix Page constructor

ikreymer committed:
fix test, update crawler image to 1.1.0-beta.1

ikreymer committed:
fix load_state -> loadState

ikreymer committed:
fix models to contain qaUsage, qaCrawlExecSeconds in org and org out

ikreymer committed:
rename FRONTEND_ALIAS -> FRONTEND_ORIGIN
add resolve_internal_access_path() to storages which prepends the frontend_origin
in get_internal_crawl_out()

tw4l commented on backend/test/test_qa.py:
```diff
@@ -0,0 +1,167 @@
+    assert completed
+
+
+def test_qa_completed(crawler_crawl_id, crawler_auth_headers, default_org_id):
```
Could add a test here to check the org stats to make sure the execution time for QA is being updated?

ikreymer replied:
Added in https://github.com/webrecorder/browsertrix-cloud/pull/1586/commits/bc3bc4427a802da2d1c40dc4af3b8446cefca3dd as a separate test.

tw4l reviewed (approved):
Looks good! Been testing on dev and locally, and went through the code again. Made a suggestion for an extra test to add but otherwise I think we're good to merge to main!

The other thing we discussed but decided to do in a follow-up is to add a check before starting QA runs that the crawl was not created with a `0.x.x` version of the crawler, and to return an appropriate status code/not start the QA run if so. We can check `Crawl.image` for this.

ikreymer committed:
add test for qaCrawlExecuted and qaUsage, ensure qaUsage is populated as well
by calling inc_crawl_complete_stats for both regular crawls and qa runs

ikreymer committed:
filterBy -> filterQABy
